### PR TITLE
feat(segment): ionChange will only emit from user committed changes

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

`ionChange` is emitted on `ion-segment` when the value is programmatically set. 

Issue URL: Internal

## What is the new behavior?

- `ionChange` is only emitted for user committed actions (selecting an `ion-segment-button`, dragging to a new active segment button, using keyboard navigation to activate an `ion-segment-button`). 

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Other information

Dev-build: `6.2.8-dev.11663602676.1e458a39`

### CODE REVIEW

File: core/src/components/segment/segment.tsx [hunk 1]
```diff
@@ -181,11 +177,21 @@ export class Segment implements ComponentInterface {
 
     const value = this.valueAfterGesture;
     if (value !== undefined) {
-      this.ionChange.emit({ value });
```
Comment (sean-perkins) [inline core/src/components/segment/segment.tsx:184]: If you compare the changes side-by-side, you might think `ionChange` is no longer emitted here. While I did remove this event emission, the earlier function calls execute `setNextIndex` → `checkButton` → `ionChange.emit`.

It was removed at this point as it was redundant and causing an extra event emission (once `ionChange.emit` was introduced to `checkButton`).

File: BREAKING.md [hunk 2]
```diff
@@ -84,6 +85,11 @@ iOS:
 |`$range-ios-knob-box-shadow`|`0 3px 1px rgba(0, 0, 0, .1), 0 4px 8px rgba(0, 0, 0, .13), 0 0 0 1px rgba(0, 0, 0, .02)`|`0px 0.5px 4px rgba(0, 0, 0, 0.12), 0px 6px 13px rgba(0, 0, 0, 0.12)`|
 |`$range-ios-knob-width`|`28px`|`26px`|
 
+<h4 id="version-7x-slides">Segment</h4>
```
Comment (farisfaisalthena) [inline BREAKING.md:88]: the `id` should be `version-7x-segment` instead of `version-7x-slides` right?

Comment (sean-perkins) [reply BREAKING.md:88]: Correct - great catch!

File: core/src/components/segment/segment.tsx (hunk 1 above)
Comment (amandaejohnston) [reply core/src/components/segment/segment.tsx:184]: I'm not seeing `ionChange` fire when dragging the segment (on Windows Chrome), only when clicking.

File: core/src/components/segment/segment.tsx [hunk 3]
```diff
@@ -76,15 +75,13 @@ export class Segment implements ComponentInterface {
...
 
   @Watch('value')
   protected valueChanged(value: string | undefined, oldValue: string | undefined | null) {
     this.ionSelect.emit({ value });
-    if (oldValue !== '' || this.didInit) {
-      if (!this.activated) {
-        this.ionChange.emit({ value });
-      } else {
+    if (oldValue !== '') {
+      if (this.activated) {
```
Comment (amandaejohnston) [inline core/src/components/segment/segment.tsx:84]:
```suggestion
    if (oldValue !== '' && this.activated) {
```

**Review by amandaejohnston — CHANGES_REQUESTED** [review]

**Comment by sean-perkins** [comment]
I need to find a better way of testing the pointer-up behavior. I'll re-ping for reviews once I figure that out.

File: core/src/components/segment/test/segment-events.e2e.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+        await firstButton.hover();
+        await page.mouse.up();
+
+        expect(ionChangeSpy).toHaveReceivedEventTimes(0);
+      });
+    });
+  });
+
+  test.describe('should not emit', () => {
```
Comment (amandaejohnston) [inline core/src/components/segment/test/segment-events.e2e.ts:121]: Optional: Might be good to make this more descriptive, something like "should not emit on button click." The descriptions here are very similar to the ones for dragging above, so it's a bit confusing at a glance.

**Review by amandaejohnston — APPROVED** [review]
Working great now, nice work on this!

File: core/src/components/segment/segment.tsx [hunk 5]
```diff
@@ -76,18 +75,15 @@ export class Segment implements ComponentInterface {
   /**
    * the value of the segment.
    */
-  @Prop({ mutable: true }) value?: string | null;
```
Comment (liamdebeasi) [inline core/src/components/segment/segment.tsx:79]: Why did we remove `null`?

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
Tabbing and selecting a button with "Enter" does not emit `ionChange` on a segment with no initial value.

File: core/src/components/segment/segment.tsx (hunk 5 above)
Comment (sean-perkins) [reply core/src/components/segment/segment.tsx:79]: After reviewing the internals of the `ion-segment` component, I could not find a reason why the `value` prop accepts `null` - as nothing from segment appears to assign a value of `null`. The change was made as part of this commit: https://github.com/ionic-team/ionic-framework/commit/ba2230510e95b31d8a6a6ab103ca09888f8237b5#diff-7e01db7218b2b21c70a8e558819e7cdb4f6b6220459509b3bd798371a6555277

There were also a few inconsistencies with the types:
-  `ionChange` emits a type of `string | undefined`, but the value type is `string | undefined | null`. 
-  The types for `newValue` and `previousValue` for the watch callback: https://github.com/ionic-team/ionic-framework/blob/main/core/src/components/segment/segment.tsx#L82

This resulted in type errors when trying to emit the `value` against `ionChange`:

e.g.:
```
const { value } = this;
this.ionChange.emit({ value });
```

> Type 'string | null | undefined' is not assignable to type 'string | undefined'.
  Type 'null' is not assignable to type 'string | undefined'.

Unless type narrowing or type casting (like the previous implementation did with the `@Watch` callback function. 

I'm open to feedback on this though - if we know a reason that `null` should be an accepted value type.

**Review by liamdebeasi — APPROVED** [review]
Great job!

File: core/src/components/segment/segment.tsx (hunk 5 above)
Comment (sean-perkins) [reply core/src/components/segment/segment.tsx:79]: After doing more research into this, I think the original intent of `null` was to offer developers a way to clear the checked item in the segment. 

I will update the breaking change docs to reflect this, but assigning a value of `''` would be preferred in this use case.

**Pull request merged**